compiler/rustc_transmute/src/nfa.rs

```diff
@@ -0,0 +1,425 @@
+        let edge = source.entry(Transition::Byte(Byte::Init(bits))).or_default();
+        edge.insert(accepting);
+
+        Nfa { transitions, start, accepting }
+    }
+
+    // Constructs an `Nfa` that describes the layout of primitive number type (e.g., `u8`, `f32`, `i64`) of
+    // `width_in_bytes` size.
+    #[allow(non_snake_case)]
+    pub fn number(width_in_bytes: usize) -> Self {
+        core::iter::repeat_with(Self::u8)
+            .take(width_in_bytes)
+            .reduce(|a, b| a.concat(b))
+            .unwrap_or_else(Self::unit)
+    }
+
+    // Constructs an `Nfa` that describes the layout of padding of `width_in_bytes` size.
+    #[allow(non_snake_case)]
+    pub fn padding(width_in_bytes: usize) -> Self {
+        core::iter::repeat_with(Self::uninit)
+            .take(width_in_bytes)
+            .reduce(|a, b| a.concat(b))
+            .unwrap_or_else(Self::unit)
+    }
+
+    pub fn from_ty(ty: Ty<'tcx>, tcx: TyCtxt<'tcx>) -> Result<Self, ()> {
+        use rustc_middle::ty::FloatTy::*;
+        use rustc_middle::ty::IntTy::*;
+        use rustc_middle::ty::TyKind;
+        use rustc_middle::ty::UintTy::*;
+        use rustc_target::abi::Align;
+        use rustc_target::abi::HasDataLayout;
+        use std::alloc::Layout;
+        use std::iter;
+
+        let target = tcx.data_layout();
+
+        match ty.kind() {
+            ty::Bool => Ok(Self::bool()),
+
+            ty::Int(I8) | ty::Uint(U8) => Ok(Self::u8()),
+            ty::Int(I16) | ty::Uint(U16) => Ok(Self::number(2)),
+            ty::Int(I32) | ty::Uint(U32) | ty::Float(F32) => Ok(Self::number(4)),
+            ty::Int(I64) | ty::Uint(U64) | ty::Float(F64) => Ok(Self::number(8)),
+            ty::Int(I128) | ty::Uint(U128) => Ok(Self::number(16)),
+            ty::Int(Isize) | ty::Uint(Usize) => Ok(Self::number(target.pointer_size.bytes_usize())),
+
+            ty::Adt(adt_def, substs_ref) => {
+                use rustc_middle::ty::AdtKind;
+                // The layout begins with this adt's visibility.
+                let vis = Self::vis(Vis::Adt(*adt_def));
+
+                // And is followed the layout(s) of its variants
+                match adt_def.adt_kind() {
+                    AdtKind::Struct => {
+                        // is the layout well-defined?
+                        if !adt_def.repr().c() {
+                            return Err(());
+                        }
+
+                        Ok(vis.concat(Self::from_repr_c_variant(
+                            ty,
+                            *adt_def,
+                            substs_ref,
+                            None,
+                            adt_def.non_enum_variant(),
+                            tcx,
+                        )?))
+                    }
+                    AdtKind::Enum => {
+                        // is the layout well-defined?
+                        if !(adt_def.repr().c() || adt_def.repr().int.is_some()) {
```

Comment:
should be `&&` here.

Reply:
Doesn't this have a well-defined layout?
```rust
#[repr(u8)]
enum Foo { A, B, C }
```

Reply:
Yes, it does.
However, `#[repr(u8)] enum Foo { A(u8,u32,u8) }` does not